aws-sdk-resources/lib/aws-sdk-resources/services/s3/file_downloader.rb

```diff
@@ -0,0 +1,163 @@
+          get_range(construct_chunks(file_size))
+        else
+          compute_mode(file_size, resp.parts_count)
+        end
+      end
+ 
+      def compute_mode(file_size, count)
+        chunk_size = compute_chunk(file_size)
+        part_size = (file_size.to_f / count.to_f).ceil
+        if chunk_size < part_size
+          get_range(construct_chunks(file_size))
+        else
+          get_part(count)
+        end
+      end
+
+      def construct_chunks(file_size)
+        offset = 0
+        default_chunk_size = compute_chunk(file_size)
+        chunks = []
+        while offset <= file_size
+          progress = offset + default_chunk_size
+          if progress < file_size
+            chunks << "bytes=#{offset}-#{progress}"
+          else
+            chunks << "bytes=#{offset}-#{file_size}"
+          end
+          offset = progress + 1
+        end
+        chunks
+      end
+
+      def compute_chunk(file_size)
+        if @chunk_size && @chunk_size > file_size
+          raise ArgumentError, ":chunk_size shouldn't exceed total file size."
+        else
+          default_chunk_size = @chunk_size || [(file_size.to_f / MAX_PARTS).ceil, MIN_CHUNK_SIZE].max.to_i
+        end
+      end
+
+      def sort_files(files)
+        # sort file by start range count or part number
+        files.sort do |a, b|
+          a[/([^\=]+)$/].split('-')[0].to_i <=> b[/([^\=]+)$/].split('-')[0].to_i
+        end
+      end
+
+      def concatenate_files(files)
+        File.open(@path, 'w', encoding: 'UTF-8') do |obj|
+          sort_files(files).each do |file|
+            File.open(file, 'r', encoding: 'UTF-8') do |f|
+              obj.write(f.read)
+            end
+            File.unlink(file)
+          end
+        end
+      end
+
+      def get_range(chunks)
+        count = 0
+        threads = []
+        file_chunks = []
+        dir = Dir.tmpdir
+        while count < chunks.size
+          remains_count = chunks.size - count
+          cnt = @thread_count < remains_count ? @thread_count : remains_count
+          cnt.times do
+            threads << Thread.new(chunks.shift) do |chunk|
+              resp = @client.get_object(bucket: @bucket, key: @key, range: chunk)
+              file_chunks << file = File.join(dir, chunk)
+              File.open(file, 'w', encoding: 'UTF-8') do |f|
+                f.write(resp.body.read)
+              end
+            end
+          end
+          threads.each(&:join)
+          count += cnt
+        end
+        concatenate_files(file_chunks)
+      end
+
+      def get_part(parts)
+        threads = []
+        file_parts = []
+        dir = Dir.tmpdir
+        parts.times do |i|
+        # partNumber starts from 1
+          part = i + 1
+          threads << Thread.new(part) do
+            resp = @client.get_object(bucket: @bucket, key: @key, part_number: part)
+            file_parts << file = File.join(dir, 'part_number=' + part.to_s)
```

Comment:
Good to know, Initially, I use `=` because for ranges, it will look like `bytes=12345-56789`, thus I use `part_number=` prefix to echo that, sounds like it might not be a good option there. Happy to discuss a better naming strategy : )

P.S. `sort_files` is introduced because I noticed that original Ruby string compare is having trouble in comparing strings like "12345-56789" in a expected numerical way.